src/azure-cli-core/azure/cli/core/parser.py

```diff
@@ -165,12 +178,96 @@ def enable_autocomplete(self):
+        def has_extension_name(command_source):
+            is_extension_command_source = isinstance(command_source, ExtensionCommandSource)
+            has_extension_name = False
+
+            if is_extension_command_source:
+                has_extension_name = hasattr(command_source, 'extension_name')
+
+            return is_extension_command_source and has_extension_name
+
+        # If the arguments have been processed into a namespace...
+        if self._namespace:
+            # Select the parsed command.
+            if hasattr(self._namespace, 'command'):
+                command = self._namespace.command
+        # Parse parameter names from user input.
+        if self._raw_arguments:
+            raw_arguments = self._raw_arguments
+            parameters = extract_safe_params(self._raw_arguments)
+
+        for parameter in parameters:
+            parameter_set.add(parameter)
+
+        # If we can retrieve the extension from the current parser's command source...
+        if has_extension_name(self.command_source):
+            extension = self.command_source.extension_name
+        # Otherwise, the command may have not been in a command group. The command source will not be
+        # set in this case.
+        elif action and action.dest == '_subcommand':
+            # Get all parsers in the set of possible actions.
+            parsers = list(action.choices.values())
+            parser = parsers[0] if parsers else None
+            # If the first parser comes from an extension...
+            if parser and has_extension_name(parser.command_source):
+                # We're looking for a subcommand under an extension command group. Set the
+                # extension to reflect this.
+                extension = parser.command_source.extension_name
+            # Extend the command if the first raw argument is not a parameter.
+            if raw_arguments and raw_arguments[0] not in parameter_set:
+                command = '{cmd} {arg}'.format(cmd=command, arg=raw_arguments[0])
+        # Otherwise, only set the extension if every subparser comes from an extension. This occurs
+        # when an unrecognized argument is passed to a command from an extension.
+        elif isinstance(self.subparser_map, dict):
+            for _, subparser in self.subparser_map.items():
+                if isinstance(subparser.command_source, ExtensionCommandSource):
+                    extension = subparser.command_source.extension_name
+                else:
+                    extension = None
+                    break
+
+        return command, parameters, extension
+
+    def _get_failure_recovery_recommendations(self, action=None, **kwargs):
+        # Gets failure recovery recommendations
+        from azure.cli.core import __version__ as core_version
+        failure_recovery_arguments = self._get_failure_recovery_arguments(action)
+        recommendations = AzCliCommandParser.recommendation_provider(core_version,
```

Comment:
Will this static method always return an empty list ? If yes, it seems that `_suggestion_msg`  will also always be empty.

Reply:
`self._suggestion_message` is also used for the spellchecker recommendations in `AzCliCommandParser._check_value`. The intention is to allow extensions to make additional failure recovery recommendations. See my [draft PR](https://github.com/Azure/azure-cli-extensions/pull/1536) for an example.

Reply:
Got it. I'm doubting that is it a good practice to inject core static function in an extension? And could you please help share an e2e case to help me understand the whole user scenario? Does it means that the `error` function's behavior will depend on the extension? If an extension didn't override this function, it will only print `recommendation_provider: version: %s, command: %s, parameters: %s, extension: %s` without other messages ?